server/implementation/src/main/java/io/smallrye/graphql/spi/datafetcher/CompletableFutureDataFetcherService.java

```diff
@@ -0,0 +1,126 @@
+        this.argumentHelper = new ArgumentHelper(operation.getArguments());
+        this.partialResultHelper = new PartialResultHelper();
+        this.batchLoaderHelper = new BatchLoaderHelper();
+    }
+
+    @Override
+    public <T> T get(DataFetchingEnvironment dfe, DataFetcherResult.Builder<Object> resultBuilder) throws Exception {
+        eventEmitter.fireBeforeDataFetch();
+
+        try {
+            Object[] transformedArguments = argumentHelper.getArguments(dfe);
+            CompletionStage<Object> futureResultFromMethodCall = reflectionHelper.invoke(transformedArguments);
+
+            return (T) futureResultFromMethodCall.handle((result, throwable) -> {
+                if (throwable instanceof CompletionException) {
+                    //Exception thrown by underlying method may be wrapped in CompletionException
+                    throwable = throwable.getCause();
+                }
+
+                if (throwable != null) {
+                    eventEmitter.fireOnDataFetchError(dfe.getExecutionId().toString(), throwable);
+                    if (throwable instanceof GraphQLException) {
+                        GraphQLException graphQLException = (GraphQLException) throwable;
+                        partialResultHelper.appendPartialResult(resultBuilder, dfe, graphQLException);
+                    } else if (throwable instanceof Exception) {
+                        throw msg.dataFetcherException(operation, throwable);
+                    } else if (throwable instanceof Error) {
+                        throw ((Error) throwable);
+                    }
+                } else {
+                    try {
+                        resultBuilder.data(fieldHelper.transformResponse(result));
+                    } catch (AbstractDataFetcherException te) {
+                        te.appendDataFetcherResult(resultBuilder, dfe);
+                    }
+                }
+
+                return resultBuilder.build();
+            });
+
+        } catch (AbstractDataFetcherException abstractDataFetcherException) {
+            //Arguments or result couldn't be transformed
+            abstractDataFetcherException.appendDataFetcherResult(resultBuilder, dfe);
+            eventEmitter.fireOnDataFetchError(dfe.getExecutionId().toString(), abstractDataFetcherException);
+        } catch (GraphQLException graphQLException) {
+            partialResultHelper.appendPartialResult(resultBuilder, dfe, graphQLException);
+            eventEmitter.fireOnDataFetchError(dfe.getExecutionId().toString(), graphQLException);
+        } catch (SecurityException | IllegalAccessException | IllegalArgumentException ex) {
+            //m.invoke failed
+            eventEmitter.fireOnDataFetchError(dfe.getExecutionId().toString(), ex);
+            throw ex;
+        } finally {
+            eventEmitter.fireAfterDataFetch();
+        }
+
+        return (T) CompletableFuture.completedFuture(resultBuilder.build());
```

Comment:
Is this really what you want? It seems that something wrong happened and you still returned a future. Shouldn't you send the failure (in a future) ?

Reply:
graphql can do partial responses, so the errors are part of a valid response.